Add support for kube cert signing requests

jandubois opened the pull request:
When the `ca` field for an ssl cert definition is empty, create a kube CSR instead.

If the `-autoApprove` commandline option is set, then also attempt to approve the request.

The secrets generator will wait until all kube CSRs have been approved. Failure to auto-approve are ignored, but any errors while checking the status of pending CSRs will terminate the generator with an error.

Add `append_kube_ca` property to CA cert options. When `true`, the secrets generator will append the `ca.crt` from the kube service account to the CA cert generated for this variable.

Allow template expansion for all variable options. Previously this was hardcoded for the `alternative_names` list only, but now is available anywhere under the `options` key.

In addition this commit adds a `--set` commandline option to specify additional template environment variables.

Unrelated, this commit also removes the `common_name` setting, which is not used anywhere in the cert generation code.

jandubois committed:
Allow template expansion for all variable options

Previously this was hardcoded for the `alternative_names` list
only, but now is available anywhere under the `options` key.

In addition this commit adds a --set commandline option to specify
additional template environment variables.

Unrelated, this commit also removes the `common_name` setting,
which is not used anywhere in the cert generation code.

[jsc#CAP-587]

jandubois committed:
Add append_kube_ca property to CA options

When true, the secrets generator will append the ca.crt from the kube
service account to the CA cert generated for this variable.

[jsc#CAP-587]

jandubois committed:
Add support for kube cert signing requests

When the `ca` field for an ssl cert definition is empty, create a kube
CSR instead.

If the -autoApprove commandline option is set, then also attempt to
approve the request.

The secrets generator will wait until all kube CSRs have been approved.
Failure to auto-approve are ignored, but any errors while checking the
status of pending CSRs will terminate the generator with an error.

[jsc#CAP-587]

mook-as commented on main.go:
```diff
@@ -57,7 +64,26 @@ var secretsName = flag.String(
+var templateEnv = make(envSettings)
+
+func (env *envSettings) String() string {
+	return ""
```
Returning an empty string could be confusing if we ever see this.

jandubois replied:
I don't know when we would see this. And I have no idea how to represent a map value here. I'm happy to put in whatever you feel appropriate.

Given that the only caller of this program is a bash script under our control too, I'm not sure it matters much.

jandubois replied:
I tried to return ` fmt.Sprintf("%v", env)`, but that shows a warning when running the tests:

```
./main.go:72:9: Sprintf format %v with arg env causes recursive String method call
```

I think returning the empty string is fine, given that the code will not be executed anyways.

mook-as replied:
Sure, or just some hard coded `<env settings>` string or something; I'm mostly thinking _empty_ string would be confusing (since we wouldn't be able to troubleshoot).

jandubois replied:
Fixed in c605f62

mook-as commented on model/manifest.go:
```diff
@@ -60,44 +65,50 @@ type CVOptions struct {
 
 // GetManifest loads a manifest from file or string
-func GetManifest(r io.Reader) (Manifest, error) {
+func GetManifest(r io.Reader, env map[string]string) (Manifest, error) {
```
We should document what `env` is for in the method doc comment, so that [the documentation](https://godoc.org/github.com/SUSE/scf-secret-generator/model#GetManifest) is reasonable.

jandubois replied:
Fixed in 8cf5338

jandubois replied:
Fixed in 8cf5338

mook-as commented on model/manifest.go:
```diff
@@ -60,44 +65,50 @@ type CVOptions struct {
+		return manifest, errors.New("'variables' section not found in manifest")
+	}
+
+	_, err = expandTemplates(raw["variables"], env)
```
I take it that we're discarding the result just to verify the expansion? If yes, document that.

If not, document what we _are_ doing… it's not modifying `raw["variables"]` in place is it? It might be better to just copy in that case…

jandubois replied:
> it's not modifying raw["variables"] in place is it? It might be better to just copy in that case…

It is doing just that. Recursively copying larger and larger tree structures in general feels like a bad idea, although in this particular instance the structures are reasonably small.

I did run into some problems when trying to implement everything via copying, but I can't remember anymore (this commit is already 6 weeks old).

I'm happy to add a comment, but not really keen to rewrite this from scratch, given that it works "as documented" (`expandTemplates` docs says that it expands "in-place").

jandubois replied:
Fixed in 8cf5338

jandubois replied:
I've documented what we are doing, so resolving this one as well. :)

mook-as commented on model/manifest.go:
```diff
@@ -106,26 +117,88 @@ func GetManifest(r io.Reader) (Manifest, error) {
+		if strings.Contains(str, "{{") {
+			t, err := template.New("").Parse(str)
+			if err != nil {
+				return nil, fmt.Errorf("Can't parse template in `%s`: %v", str, err)
```
It would be good to document the path taken to get to this string.

jandubois replied:
Yeah, but where do I get the path?

I expect this error to just catch malformed template expressions, and the error should include the full template. So it should be easy enough to find this in the rolemanifest.

mook-as replied:
We _could_ pass in the path as we recurse through the object. But it's probably fine for now… it's just annoying later :)

mook-as commented on secrets/secrets.go:
```diff
@@ -139,6 +142,15 @@ func (sg *SecretGenerator) getSecretInterface() (secretInterface, error) {
 }
 
+// getSecretInterface returns a secrets interface for the namespace
+func (sg *SecretGenerator) getCertificateSigningRequestInterface() (util.CertificateSigningRequestInterface, error) {
```
The documentation doesn't match the function name. (And the function name is rather a mouthful.)

jandubois replied:
Fixed in 8cf5338.

The name is indeed rather long, but it follows the same name scheme used for 2 other functions (and interfaces), so I decided to leave it as-is.

mook-as commented on ssl/ssl.go:
```diff
@@ -137,13 +149,43 @@ func GenerateCerts(certInfo map[string]CertInfo, namespace, clusterDomain string
+	var err error
 	if len(errs) > 0 {
-		return errs[0]
+		err = errs[0]
```
Is there a point to attempting to sign the certs when stuff is known to be broken? So we can go delete the CSRs? If yes, it may be clearer to do that with `defer` (but then it might make the error reporting harder…)

jandubois replied:
Yes, this exists only to make sure we delete the outstanding CSRs.

I don't think using `defer` will make this clearer, because it interferes with proper error handling, as you already noted.

However, this now reminds me that I wanted to attempt do delete the CSR before creating it, just in case it somehow escaped proper cleanup in an earlier run.

mook-as commented on ssl/ssl.go:
```diff
@@ -265,18 +321,80 @@ func createCert(certInfo map[string]CertInfo, namespace, clusterDomain string, s
+				certificates.UsageClientAuth,
+				certificates.UsageServerAuth,
+			},
+			// expiration is determined by apiserver and cannot be configured
```
Do we need to figure out the procedure to renew the certs? (Probably as a new task)

jandubois replied:
The procedure should be the same as for any other certs we generate: secrets rotation.

I would hope that maybe Stratos will eventually have a way to warn an admin about certs that are going to expire soon. Alternatively the admin could run something like `havener` to quickly get the current state.

mook-as replied:
Okay, we _do_ have it [documented](https://susedoc.github.io/doc-cap/develop/cap-guides/html/cha-cap-certificates.html#sec-cap-secrets-rotation), so we should be fine.

mook-as commented on ssl/ssl.go:
```diff
@@ -265,18 +321,80 @@ func createCert(certInfo map[string]CertInfo, namespace, clusterDomain string, s
+		}
+	}
+	cond := csr.Status.Conditions[0]
+	if cond.Type != certificates.CertificateApproved {
```
The only possible values here are approved or denied (there are no unrelated conditions in this array), right? We can't have the approval condition later in the condition list?

jandubois replied:
Technically it would be possible to have multiple `Conditions` in the slice, but I don't see how that makes sense.

I guess it really depends on the cert approver (you could have a custom implementation), but given that there are only 2 defined type values, I don't know what we should do. Loop through all conditions and stop at the first approved or denied type?

mook-as replied:
That was my thought, yes.

jandubois replied:
Fixed in c605f62

mook-as commented on ssl/ssl_test.go:
```diff
@@ -315,6 +398,27 @@ func TestCreateCA(t *testing.T) {
 
 		assert.NotEqual(t, secrets.Data[certInfo[certID].PrivateKeyName], []byte{})
 		assert.NotEqual(t, secrets.Data[certInfo[certID].CertificateName], []byte{})
+		assert.Regexp(t, "(?s)\\A-----BEGIN CERTIFICATE-----\\n.*\\n-----END CERTIFICATE-----\\n\\z",
```
Use backticks for the string delimiter to skip the need to escape backslashes.  Too bad there isn't `assert.HasPrefix` or something like that…

jandubois replied:
Fixed in 8cf5338

mook-as commented on ssl/ssl_test.go:
```diff
@@ -315,6 +398,27 @@ func TestCreateCA(t *testing.T) {
+		assert.NotEqual(t, secrets.Data[certInfo[certID].PrivateKeyName], []byte{})
+		assert.NotEqual(t, secrets.Data[certInfo[certID].CertificateName], []byte{})
+		assert.Contains(t, string(secrets.Data[certInfo[certID].CertificateName]),
+			"-----END CERTIFICATE-----\n-----BEGIN CERTIFICATE-----")
```
Shouldn't we test that the content of `kubeca.crt` was inserted (and not, say, two copies of the internal CA)?

jandubois replied:
Fixed in 8cf5338

mook-as reviewed (changes requested):
Mostly fine, just some localized changes.

jandubois committed:
Code review feedback; mostly improved comments

Functional change: Always try to delete a kube csr before creating
it, just in case there are left-over requests from a failed previous
run.

[jsc#CAP-587]

jandubois committed:
Review feedback: allow unknown certificate request condition types

Also don't use empty string to serialize envSettings object.

[jsc#CAP-587]
